### ISSUE TITLE
Gradients of local displacement energy 

### ISSUE THREAD

**Issue by bashkanov** [YipengHu/label-reg#18]
Dear @YipengHu,

thank you for sharing this code with us! 
I'm trying to understand your `local_displacement_energy` function. In your paper you refer to (_Rueckert et al. 1999_) where bending energy is clearly defined. Implementation of L1, L2 gradient norms are also pretty straightforward. But I still cannot get the intuition behind your `gradient_dx/y/z` functions. 

It's clear to me why gradient_dx/y/z returns the vol shape less by 2 for x,y,z, axes, e.g. the input shape `[batch, 90, 90, 90, 3]` will result into shape `[batch, 88, 88, 88, 3]` for _dx, dy, dz._ If we want to preserve the original shape we should append zeroes accordingly as it states in this tf.doc [image_gradients](https://www.tensorflow.org/api_docs/python/tf/image/image_gradients): 
> Both output tensors have the same shape as the input: [batch_size, h, w, d]. The gradient values are organized so that [I(x+1, y) - I(x, y)] is in location (x, y). That means that _dy_ will always have zeros in the last row, and _dx_ will always have zeros in the last column.

- Could you please elaborate why do you prefer to lose some information, instead of appending zeroes to preserve the shape?
- Why do we need to divide the gradients by two?

I’m eager to receive your response!

**Comment by YipengHu** [comment]
sorry @bashkanov  i overlooked your ticket un till closed it just now :p Do you still have the questions or it is solved now?

**Comment by bashkanov** [comment]
The implementation part is clear to me now :+1:.  Could you please elaborate on advantageous of calculating the image gradient with a step size of two (as you did) instead of taking it for neighboring pixels? Maybe there is a paper explaining it well?

> sorry @bashkanov i overlooked your ticket un till closed it just now :p Do you still have the questions or it is solved now?

**Comment by YipengHu** [comment]
There is nothing fancy - this is standard gradient calculation using finite difference. You can use more than one pixels in each direction if not for computational efficiency. You can also convolve with derivatives of Gaussian using efficient FFT. The implemented is only one of the numerical solutions for computing gradient.

**Comment by bashkanov** [comment]
Thank you for answering my question to satisfy my curiosity!
